[docs] 2/n add airflow migration strategies: TaskFlow, Kubernetes (#22614)
dagster-io/dagster: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx

        get_logs=True,
    )
```

Now, let’s show how the same image would be run on Kubernetes using Dagster pipes and the `dagster_k8s` wrapper.

```python
from dagster import AssetExecutionContext, asset
from dagster_k8s import PipesK8sClient

@asset
def k8s_pipes_asset(context: AssetExecutionContext, k8s_pipes_client: PipesK8sClient):
    return k8s_pipes_client.run(
        context=context,
        image="example-data-pipeline:latest",
        base_pod_spec={
            "containers": [
                {
                    "name": "data-processing-rs",
                    "image": "data-processing-rs",
                }
            ]
        },
    ).get_materialize_result()
```

The primary difference between Airflow and Dagster are how the k8s pod specification are expose. In Airflow, they are passed as parameters the `KubernetesPodOperator`, whereas in Dagster they are passed as a `base_pod_spec` dictionary to the `k8s_pipes_client.run` method. Another difference is that in Airflow, `get_logs` is required to capture _stdout_, however, with Dagster they are automatically captured to on the _stdout_ tab of the step output.

### Advanced pipes usage

> In the above example, we demonstrated how to run images on Kubernetes using the `dagsater_k8s` library. One of the biggest benefits of Dagster Pipes, however, is that you can leverage the `dagster_pipes` library from within your containerized pipeline to access the full Dagster context, and emit events back to the Dagster UI.

Reviewer: ```suggestion
In the above example, we demonstrated how to run images on Kubernetes using the `dagster_k8s` library. One of the biggest benefits of Dagster Pipes, however, is that you can leverage the `dagster_pipes` library from within your containerized pipeline to access the full Dagster context, and emit events back to the Dagster UI.
```

> # Migrating Airflow to Dagster

Author: How do we feel about the structure of this page - do we need an index at the top? Should it be split into multiple pages?

We will continue to append new migration strategies / mapping of concepts over time.

<Note>
  Looking for an example of an Airflow to Dagster migration? Check out the{" "}

>   <a href="https://github.com/dagster-io/dagster-airflow-migration-example">

Author: Reminder to archive this repository as we deprecate `dagster-airflow`

Author: Repository has been archived.

## Translating Airflow pipelines to Dagster

If your organization makes heavy use of the `PythonOperator`, the Airflow TaskFlow API, or the `KubernetesPodOperator`, then migrating to Dagster may be relatively simple!

### Migrating pipelines that use the TaskFlow API

Arguably, the Airflow piplines that are the most simple to migrate to Dagster are those that make use of the [TaskFlow](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/taskflow.html) API paradigm. In this paradigm, pipelines are constructed using Airflow `@task` decorators that can be easily mapped to the Dagster `@op` decorator. Let’s walk through the translation of the `tutorial_taskflow_api.py` example DAG.

```python
import json

import pendulum

from airflow.decorators import dag, task

@dag(
    schedule=None,
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    catchup=False,
    tags=["example"],
)

> def tutorial_taskflow_api():

Author: Ok, I've removed the `@op` example and added smaller snippets of the task and asset before showing the full examples.

    order_data_dict = json.loads(data_string)
    return order_data_dict

@op()
def transform(order_data_dict: dict):
    """Transform task

    A simple Transform task which takes in the collection of order data and computes the total order
    value.
    """
    total_order_value = 0

    for value in order_data_dict.values():
        total_order_value += value

    return total_order_value

@op()
def load(total_order_value: float):
    """Load task

    A simple Load task which takes in the result of the Transform task and instead of saving it to
    end user review, just prints it out.
    """

    print(f"Total order value is: {total_order_value:.2f}")

@job
def airflow_taskflow_example():
    op1 = extract()
    op2 = transform(op1)
    load(op2)

defs = Definitions(
    jobs=[airflow_taskflow_example]
)
```

Alternatively, to extract the full power of Dagster, you may want to consider leveraging _asset_ paradigm. In this example, we will consolidate the individual operations into a single asset that represents the resulting data of our pipeline.

```python
import json

from dagster import Definitions, MaterializeResult, define_asset_job, asset

@asset

> def extract_transform_load_asset():

Author: Would like feedback as to whether or not we want to show both the `@op` approach and the consolidated asset.

Originally I had done three assets: extract, transform, and load -- would that be preferred here?

Reviewer: I would choose any asset variation over the standalone `@op` approach. Using `@op` in a graph-backed asset could be a nice middle ground. 

Having three assets also makes sense, as users may want a MaterializeResult for each of them.

Author: For now I've removed the `@op` example and replaced it with a 1-to-1 mapping of `@asset`. This is using the implicit asset outs to keep it simple, as I agree that introducing graph-backed right away may be overwhelming.

Reviewer 2: Let's focus on asset!

```python
from airflow import DAG
from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator
from pendulum import datetime

with DAG(dag_id="example_kubernetes_dag", schedule_interval=None, catchup=False) as dag:
    KubernetesPodOperator(
        image="example-data-pipeline:latest",
        name="example-kubernetes-task",
        task_id="example-kubernetes-task",
        get_logs=True,
    )
```

Now, let’s show how the same image would be run on Kubernetes using Dagster pipes and the `dagster_k8s` wrapper.

```python
from dagster import AssetExecutionContext, asset
from dagster_k8s import PipesK8sClient

@asset
def k8s_pipes_asset(context: AssetExecutionContext, k8s_pipes_client: PipesK8sClient):
    return k8s_pipes_client.run(
        context=context,
        image="example-data-pipeline:latest",
        base_pod_spec={
            "containers": [
                {
                    "name": "data-processing-rs",
                    "image": "data-processing-rs",
                }
            ]
        },
    ).get_materialize_result()
```

> The primary difference between Airflow and Dagster are how the k8s pod specification are expose. In Airflow, they are passed as parameters the `KubernetesPodOperator`, whereas in Dagster they are passed as a `base_pod_spec` dictionary to the `k8s_pipes_client.run` method. Another difference is that in Airflow, `get_logs` is required to capture _stdout_, however, with Dagster they are automatically captured to on the _stdout_ tab of the step output.

Reviewer 3: ```suggestion
The primary difference between Airflow and Dagster are how the k8s pod specifications are exposed. In Airflow, they are passed as parameters to the `KubernetesPodOperator`, whereas in Dagster they are passed as a `base_pod_spec` dictionary to the `k8s_pipes_client.run` method. Additionally, in Airflow, `get_logs` is required to capture `stdout`. In Dagster, however, they are automatically captured on the `stdout` tab of the step output.
```

For the `stdout` tab, are you referring to the tab in the Run Details page of the UI?

Author: Yes, that's right.

> **Advanced pipes usage**

Reviewer 3: This should be a subheading, but I'm not sure what level is should be at the moment. We can fix it once we figure out what the structure of this page is going to look like.

> In the above example, we demonstrated how to easily run images on Kubernetes using the `dagsater_k8s` library. One of the biggest benefits of pipes, that we've not yet covered, is that you can leverage the `dagster_pipes` library from within your containerized pipeline to access the full Dagster context, and emit events back to the Dagster UI!

Reviewer 3: ```suggestion
In the above example, we demonstrated how to run images on Kubernetes using the `dagsater_k8s` library. One of the biggest benefits of Dagster Pipes, however, is that you can leverage the `dagster_pipes` library from within your containerized pipeline to access the full Dagster context, and emit events back to the Dagster UI.
```

Some suggestions for improving readability

> A common pattern when building containerized pipelines is to accept a large number of command-line arguments using libraries like `argparse`. However, with Dagster it is possible pass a dictionary of parameters on the Dagster context using the `extras` parameter. Then, in your pipeline code you can access the context `PipesContext.get()` if you are using Python.

Reviewer 3: ```suggestion
A common pattern when building containerized pipelines is to accept a large number of command-line arguments using libraries like `argparse`. However, with Dagster you can pass a dictionary of parameters on the Dagster context using the `extras` parameter. Then, in your pipeline code, you can access the context `PipesContext.get()` if you are using Python.
```

<Note>
  Support for Pipes clients in other programming languages are in the works, if
  you have any specific requirements, please let us know!

> </Note>

Reviewer 3: Why include this here?

Author: Removed

> For a full tutorial on how to use Dagster Pipes, please reference [Dagster Pipes tutorial](https://docs.dagster.io/concepts/dagster-pipes/subprocess).

Reviewer 3: ```suggestion
For a step-by-step walkthrough of using Dagster Pipes, refer to the [Dagster Pipes tutorial](https://docs.dagster.io/concepts/dagster-pipes/subprocess).
```

Kinda nitpicky, but using `please` in tech docs isn't recommended
